README.md
[![Build Status](https://travis-ci.org/TGuillerme/Inapp.svg?branch=master)](https://travis-ci.org/TGuillerme/Inapp)
[![Project Status: Active - The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![codecov](https://codecov.io/gh/TGuillerme/Inapp/branch/master/graph/badge.svg)](https://codecov.io/gh/TGuillerme/Inapp)
[![DOI](https://zenodo.org/badge/84838115.svg)](https://zenodo.org/badge/latestdoi/84838115)

**Inapp**: ancestral reconstruction and step counting for discrete characters with inapplicable data.

<a href="https://figshare.com/articles/Guillerme_Evolution2017_pdf/5140222"><img src="http://tguillerme.github.io/images/logo-FS.png" height="15" widht="15"/></a> 
Check out the [presentation](https://figshare.com/articles/Guillerme_Evolution2017_pdf/5140222). 

This package allows to run the algorithm described in [this paper](https://dx.doi.org/10.1093/sysbio/syy083/5238046).
You can either directly run the package on your web browser (Graphical User Interface mode!), or install it in `R` as a proper package (or both!).

The `C` implementation of the algorithm is available in [morphylib](https://github.com/mbrazeau/morphylib/) and the algorithm is implemented in the '[TreeSearch](https://github.com/ms609/TreeSearch/)' `R` package.

## Running Inapp online
Super easy! Simply click on the following link:

<a href="https://tguillerme.shinyapps.io/inapp/"><img src="http://tguillerme.github.io/images/shiny.png" height="100" widht="100"/></a> 

[https://tguillerme.shinyapps.io/inapp/](https://tguillerme.shinyapps.io/inapp/)



## Running Inapp locally on your web browser
Not hard! In `R`, simply copy/paste the following:
```r
if(!require(devtools)) install.packages("devtools")
if(!require(shiny)) install.packages("shiny")
runGitHub("Inapp", "TGuillerme")
```


## Running Inapp in your R console
Not much more complicated! In `R`, copy/paste the following:
```r
## Installing the App:
if(!require(devtools)) install.packages("devtools")
devtools::install_github("TGuillerme/Inapp")

## Running the App:
Inapp::runInapp()
```

Check out the package [vignette](https://tguillerme.github.io/inapp.html) for (many) more details on the GUI possibilities.


Authors
-------
[Thomas Guillerme](http://tguillerme.github.io), [Martin Brazeau](http://www.imperial.ac.uk/people/m.brazeau), [Martin Smith](https://community.dur.ac.uk/martin.smith/)


Citations
-------
If you are using this package, please cite both the published description of this algorithm:

* Brazeau MD, Guillerme T, Smith MR. (**2019**) An algorithm for morphological phylogenetic analysis with inapplicable data. *Systematic Biology* 68(4): 619--631. [doi:10.1093/sysbio/syy083](https://dx.doi.org/10.1093/sysbio/syy083/5238046)
    ###### [BibTeX](https://academic.oup.com/Citation/Download?resourceId=5238046&resourceType=3&citationFormat=2), [RIS](https://academic.oup.com/Citation/Download?resourceId=5238046&resourceType=3&citationFormat=0), [EndNote](https://academic.oup.com/Citation/Download?resourceId=5238046&resourceType=3&citationFormat=1), [more...](https://academic.oup.com/sysbio/article/68/4/619/5238046#)

And the DOI of this package:

 * Guillerme T, Brazeau MD, Smith MR. (**2018**). Inapp: Reconstruction of Inapplicable Discrete Characters on Phylogenetic Trees. *Zenodo*. [doi:10.5281/zenodo.1484656](http://doi.org/10.5281/zenodo.1484656)
    ###### [BibTeX](https://zenodo.org/record/1484656/export/hx), [CSL](https://zenodo.org/record/1484656/export/csl), [DataCite](https://zenodo.org/record/1484656/export/dcite3), [Dublin core](https://zenodo.org/record/1484656/export/xd), [Mendeley](https://www.mendeley.com/import/?url=https://zenodo.org/record/1484656), [more...](https://zenodo.org/record/1484656/#.XTpLtlBS8W8)


NEWS.md
Patch notes
----
Inapp v0.5.2 (2020/04/30)
=========================

### MINOR IMPROVEMENTS

  * Support characters where all tokens are inapplicable
  * Update citation information
  * Updated package for `R 4.0.0`

Inapp v0.5.1 (2019/05/08)
=========================

### MINOR IMPROVEMENTS

  * Updated package to R version 3.6.0

Inapp v0.5 (2019/03/04)
=========================

### NEW FEATURES
 
  * `Inapp` can now read any user input tree files (newick or nexus with one or multiple topologies).

### MINOR IMPROVEMENTS

  * Some error messages have been improved

Inapp v0.4.1 (2018/11/13)
=========================
  
### NEW FEATURES
  
  * Graphical options greatly improved
  * Test coverage greatly improved

### MINOR IMPROVEMENTS

  * Corrected various typos in the vignette

Inapp v0.3 (2018/01/19)
=========================
  
### MINOR IMPROVEMENTS

  * Updated references
  * Updated error messages
  * Changed mentions of "score" to "length" (in code and manuals)

Inapp v0.2.2 (2017/09/20)
=========================
  
### NEW FEATURES
  * Updated display management

### BUG FIXES
  * Many minor bug fixes

Inapp v0.2 (2017/05/01)
=========================
  
### NEW FEATURES
  * Added polymorphic characters management
  * Added `$length` element to `NA.matrix` objects (is generated from `apply.reconstruction`)

Inapp v0.1 (2017/04/27)
=========================

  * Passing first version


ui.R
shinyUI(fluidPage(
  theme = 'slimline.css',

  wellPanel(

    titlePanel("Inapplicable data reconstruction"),
    p("Brazeau, M. D., Guillerme, T., and Smith, M. R. (2019). An algorithm for Morphological Phylogenetic Analysis with Inapplicable Data. Systematic Biology 68 (4), 619-631. doi:", a(href="https://dx.doi.org/10.1093/sysbio/syy083", "10.1093/sysbio/syy083"), "."),
    hr(),

    fluidRow(

        ## ---------------
        ## Tree parameters
        ## ---------------
        column(width = 4,
          ## Tree input - input$tree
          radioButtons("tree", label = h3("Tree input method"), choices = list("Random" = 1, "User (Newick)" = 2, "Nexus input" = 3), selected = 1),

          ## Random trees
          conditionalPanel(condition = "input.tree == 1",
            ## Type of tree - input$tree_type
            selectInput("tree_type", label = "Tree topology type", choices = list("Random", "Balanced", "Left", "Right", "Left-Right"), selected = "Random"),
            ## Number of taxa - input$n_taxa
            sliderInput("n_taxa", label = "Number of taxa:", min = 3, max = 100, value = 12),
            ## Messages for tree type :
            ## Help text for balanced trees
            conditionalPanel(condition = "input.tree_type == \"Balanced\"", helpText("For a balanced tree, the number of taxa must be a power of 2 otherwise a random tree is used instead.")),
            ## Help text for balanced trees
            conditionalPanel(condition = "input.tree_type == \"Left-Right\"", helpText("For a 'Left-Right' tree, the number of taxa must be even otherwise a random tree is used instead."))
          ),

          ## Newick trees
          conditionalPanel(condition = "input.tree == 2",
            ## Enter some newick string - input$newick_tree
            textInput("newick_tree", label = h5("Enter a newick tree:"), value = "((a,b),(c,d));"),
            ## Help text for the newick trees
            helpText("Make sure the number of opening and closing brackets match and that no commas (',') or the semi-colon (';') was omitted.")
          ),

          ## Nexus trees
          conditionalPanel(condition = "input.tree == 3",
            ## Upload some nexus tree
            fileInput("nexus_tree", label = h5("Select a newick or nexus format tree")),
            ## Tree number input - input$tree_num
            numericInput("tree_num", label = h5("Selected tree:"), value = 1, min = 1)
          ),

          ## Tips and nodes options
          checkboxGroupInput("showlabels", label = "Label:",
                             choices = list("Tips" = 1, "Nodes" = 2),
                             inline=TRUE, selected = NULL)
        ),

        ## --------------------
        ## Character parameters
        ## --------------------
        column(width = 4,
          ## Character input - input$character
          radioButtons("character", label = h3("Character input method"),
                       choices = list("Random" = 1, "User" = 2, "Nexus input" = 3),
                       selected = 1),

          ## Manual character input
          conditionalPanel(condition = "input.character == 2",
            ## Character string - input$character_string
            textInput("character_string", label = h5("Enter a character string:"), value = "0?1-"),
            ## Help text for the character string
            helpText("Accepted character states are any values from 0 to 9 as well ",
                     "as - for the inapplicable token and ? for all states (missing ",
                     "data). Polymorphic characters can be entered as {01}."),
            ## Match character
            checkboxInput("matchtipchar", label = "Match character to tips",
                          value = FALSE),
            helpText("Tick to match left-right state input order state to the ",
                     "alphanumeric order of the tip labels.")
          ),

          ## Nexus character
          conditionalPanel(condition = "input.character == 3",
            ## Nexus matrix input - input$nexus_matrix
            fileInput("nexus_matrix", label = h5("Select a nexus format matrix")),
            ## Character number input - input$character_num
            numericInput("character_num", label = h5("Selected character:"), value = 1, min = 1)
          ),

          ## Colour states
          checkboxInput("colour_states", label = "Colour states", value = TRUE)

        ),

        ## -------
        ## Display
        ## -------
        column(width = 4,
          ## Reconstruction - input$method
          radioButtons("method", label = h3("Reconstruction method"), choices = list("Inapplicable Fitch" = 1, "Normal Fitch" = 2), selected = 1, inline = TRUE),

          ## Inapplicable algorithm
          conditionalPanel(condition = "input.method == 1",
            ## Which passes to print
            checkboxGroupInput("showPassInapp", label = "Show passes:",
                               choices = list("1st Downpass" = 1, "1st Uppass" = 2,
                                              "2nd Downpass" = 3, "2nd Uppass" = 4),
                               selected = c(1,2,3,4))
          ),

          ## Fitch algorithm
          conditionalPanel(condition = "input.method == 2",
            ## How to treat missing data
            radioButtons("fitch_inapp", label = h5("Inapplicable tokens are:"), choices = list("Missing data (?)" = 1, "An extra state" = 2), selected = 1, inline = TRUE),
            helpText("When treated as ?, - is equal to all states; when treated as an extra state, - is equal to a new character state."),
            ## Which passes to show?
            checkboxGroupInput("showPassFitch", label = h5("Show passes"),  choices = list("1st Downpass" = 1, "1st Uppass" = 2), selected = c(1,2))
          ),

          ## Show activations/counts - input$counts
          checkboxGroupInput("counts", label = "Show counts of:",  choices = list("Applicable regions" = 1, "State changes" = 2), selected = c(1, 2)),

          hr(),
          ## Refresh button - input$refresh
          actionButton("refresh", label = "Refresh")

        )
    )
  ),

  ## Displaying the results
  fluidRow(
    ## Plots the algorithm results
    uiOutput("plot.ui")
  ),

  ## Plot configuration
  fluidRow(
    column(width = 4,
      sliderInput("plot_height", 'Output height', 20, 2000, 600, post = 'px')
    ),
    column(width = 4,
      sliderInput("font_size", "Text size", 0.4, 2, 1)
    ),
    column(width = 4,
      ## Export format - input$export_type
      selectInput("output_type", "Export results as:",
                  choices = c("csv", "pdf", "newick", "nexus", "C-test")),
      downloadButton('downloadData', 'Download')
    ),

    conditionalPanel(condition = "input.output_type == \"C-test\"",
                     ## Enter some traversal order - input$traversal_order
                     textInput("traversal_order", label = h5("Enter an optional traversal order for output:"), value = ""),
                     ## Help text for the traversal
                     helpText("The traversal must be written in a comma separated format (e.g. 1,2,3). If left as blank the order is: tips, root, right left nodes (traversal).")
    )

  )

))



